File: 78 Sub Sets/main.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
    public:
        vector<vector<int>> subSets(int arr[], int n) {
            vector<vector<int>> result = {};
            vector<int> ans = {};
            int i = 0;
            sets(arr, ans, i, n, result);
            return result;
        }
        void sets(int arr[], vector<int> &ans, int i, int n, vector<vector<int>> &result) {
            if(n == i) {
                result.push_back(ans);
                return;
            }
            ans.push_back(arr[i]);
            sets(arr, ans, i + 1, n, result);
            ans.pop_back();
            sets(arr, ans, i + 1, n, result);
        }
        
};

int main() {

    int arr[] = {1, 2, 3};
    int n = sizeof(arr) / sizeof(arr[0]);

    Solution obj;
    vector<vector<int>> result = obj.subSets(arr, n);

    for(auto vec : result) {
        cout << "[ ";
        for(int num : vec) {
            cout << num << " ";
        }
        cout << ']';
        cout << endl;
    }

    return 0;
}

File: 78 Sub Sets/main.py
class Solution:
    def subSets(self, arr):
        result = []
        def sets(arr, res, i):
            if i == len(arr):
                result.append(res[:])
                return
            res.append(arr[i])
            sets(arr, res, i + 1)
            res.pop()
            sets(arr, res, i + 1)
        sets(arr, [], 0)
        return result

class Main:
    arr = [1, 2, 3]
    obj = Solution()
    res = obj.subSets(arr)
    print(res)